File: go/Convert.go
package main

import (
	"bufio"
	"encoding/xml"
	"fmt"
	"os"
	"strings"
)

type Family struct {
	XMLName xml.Name `xml:"family"`
	Name    string   `xml:"name"`
	Year    string   `xml:"born"`
	Address *Address `xml:"address"`
	Phone   *Phone   `xml:"phone"`
}

type Phone struct {
	XMLName        xml.Name `xml:"phone"`
	MobileNumber   string   `xml:"mobile"`
	LandlineNumber string   `xml:"landline"`
}

type Address struct {
	XMLName xml.Name `xml:"address"`
	Street  string   `xml:"street"`
	City    string   `xml:"city"`
	Postal  string   `xml:"postal"`
}

type People struct {
	XMLName xml.Name `xml:"people"`
	Persons []Person
}

type Person struct {
	XMLName   xml.Name `xml:"person"`
	FirstName string   `xml:"firstname"`
	LastName  string   `xml:"lastname"`
	Phone     *Phone   `xml:"phone"`
	Address   *Address `xml:"address"`
	Family    []Family
}

type Parser struct {
	text   *string
	pos    int
	line   int
	column int
}

func (r *Parser) peek() byte {
	// Returns the byte at the current position in the parser
	if r.pos < len(*r.text) {
		return (*r.text)[r.pos]
	}
	return 0
}

func (r *Parser) adv() {
	// Advances the position of the parser
	r.column++
	if (*r.text)[r.pos] == '\n' {
		r.line++
		r.column = 1
	}
	r.pos++
}

func makeError(err string, parser Parser) error {
	// Augments an error message with the current line number
	return fmt.Errorf("Line %v, col %v: %v", parser.line, parser.column, err)
}

func acceptChar(byt byte, parser *Parser) bool {
	// Advances the parser position if the given character is at the current position
	if parser.peek() != byt {
		return false
	}
	parser.adv()
	return true
}

func acceptWhiteSpace(parser Parser) Parser {
	// Advances the parser position until the next non-whitechar character
	for acceptChar('\t', &parser) || acceptChar('\r', &parser) || acceptChar('\n', &parser) || acceptChar(' ', &parser) {
	}
	return parser
}

func acceptPipeVal(parser Parser) (string, Parser, error) {
	// Accepts a string of type "|<string>" until endline or EOF and returns the string
	if !acceptChar('|', &parser) {
		return "", parser, makeError("Expected '|'", parser)
	}

	var s strings.Builder
	for {
		c := parser.peek()
		if strings.Contains("|\r\n\000", string(c)) { // TODO: slow, make better contains function
			return s.String(), parser, nil
		}
		s.WriteByte(c)
		parser.adv()
	}
}

func parseP(parser Parser) (string, string, Parser, error) {
	// Parses a string P|<string>|<string>
	char := acceptChar('P', &parser)
	if !char {
		return "", "", parser, makeError("Expected P", parser)
	}
	firstName, newParser, err := acceptPipeVal(parser)
	if err != nil {
		return "", "", newParser, makeError("Expected first name", newParser)
	}
	lastName, newParser, err := acceptPipeVal(newParser)
	if err != nil {
		return "", "", newParser, makeError("Expected last name", newParser)
	}
	return firstName, lastName, newParser, nil
}

func parseA(parser Parser) (*Address, Parser, error) {
	// Parses a string A|<string>|<string>|<string>
	char := acceptChar('A', &parser)
	if !char {
		return nil, parser, makeError("Expected A", parser)
	}
	street, newParser, err := acceptPipeVal(parser)
	if err != nil {
		return nil, newParser, makeError("Expected street name", newParser)
	}
	city, newParser, err := acceptPipeVal(newParser)
	if err != nil {
		return nil, newParser, makeError("Expected city", newParser)
	}
	postal, newParser, err := acceptPipeVal(newParser)
	if err != nil {
		return nil, newParser, makeError("Expected postal number", newParser)
	}
	return &Address{Street: street, City: city, Postal: postal}, newParser, nil
}

func parseF(parser Parser) (*Family, Parser, error) {
	// Parses a string F|<string>|<string>
	char := acceptChar('F', &parser)
	if !char {
		return nil, parser, makeError("Expected F", parser)
	}
	name, newParser, err := acceptPipeVal(parser)
	if err != nil {
		return nil, newParser, makeError("Expected name", newParser)
	}
	year, newParser, err := acceptPipeVal(newParser)
	if err != nil {
		return nil, newParser, makeError("Expected year", newParser)
	}
	return &Family{Name: name, Year: year}, newParser, nil
}

func parseT(parser Parser) (*Phone, Parser, error) {
	// Parses a string T|<string>|<string>
	char := acceptChar('T', &parser)
	if !char {
		return nil, parser, makeError("Expected T", parser)
	}
	mobile, newParser, err := acceptPipeVal(parser)
	if err != nil {
		return nil, newParser, makeError("Expected mobile number", newParser)
	}
	landline, newParser, err := acceptPipeVal(newParser)
	if err != nil {
		return nil, newParser, makeError("Expected landline number", newParser)
	}
	return &Phone{MobileNumber: mobile, LandlineNumber: landline}, newParser, nil
}

func parseFamily(parser Parser) (*Family, Parser, error) {
	// Parses the F of a P
	family, newParser, err := parseF(parser)
	if err != nil {
		return nil, parser, err
	}

	for true {
		newParser = acceptWhiteSpace(newParser)

		if newParser.peek() == 'A' {
			address, newParserA, err := parseA(newParser)
			if err != nil {
				return nil, parser, err
			}
			if family.Address != nil {
				// Here we read something that was already assigned which means we have a P
				// that assigns his own A after that of his F
				return family, newParser, nil
			}
			family.Address = address
			newParser = newParserA
		} else if newParser.peek() == 'T' {
			// This case is similar to A
			phone, newParserT, err := parseT(newParser)
			if err != nil {
				return nil, parser, err
			}
			if family.Phone != nil {
				return family, newParser, nil
			}
			family.Phone = phone
			newParser = newParserT
		} else {
			return family, newParser, nil
		}
	}
	return nil, Parser{}, nil
}

func parsePerson(parser Parser) (*Person, Parser, error) {
	// Parse information belonging to a Person - A 'P' and lines below it
	var person Person

	first, last, newParser, err := parseP(parser)
	if err != nil {
		return nil, parser, err
	}
	person.FirstName = first
	person.LastName = last

	for true {
		newParser = acceptWhiteSpace(newParser)
		if newParser.peek() == 'A' {
			address, newParserA, err := parseA(newParser)
			newParser = newParserA
			if err != nil {
				return nil, parser, err
			}
			person.Address = address
		} else if newParser.peek() == 'F' {
			family, newParserF, err := parseFamily(newParser)
			newParser = newParserF
			if err != nil {
				return nil, parser, err
			}
			person.Family = append(person.Family, *family)
		} else if newParser.peek() == 'T' {
			phone, newParserT, err := parseT(newParser)
			newParser = newParserT
			if err != nil {
				return nil, parser, err
			}
			person.Phone = phone
		} else {
			return &person, newParser, nil
		}
	}
	return nil, Parser{}, nil
}

func parsePeople(parser Parser) (*People, error) {
	// Parses a list of people
	var people People
	for true {
		parser = acceptWhiteSpace(parser)

		c := parser.peek()
		if c == '\000' {
			break
		}

		person, newParser, err := parsePerson(parser)
		parser = newParser
		if err != nil {
			return nil, err
		}
		people.Persons = append(people.Persons, *person)

	}
	return &people, nil
}

func stringToPeople(input string) (*People, error) {
	// Turns the input format into our intermediary data structure
	parser := Parser{&input, 0, 1, 1}
	return parsePeople(parser)
}

func peopleToXml(people People) []byte {
	// Turns our intermediary data structure to XML (we separated these into
	// functions so we can easily test them)
	out, _ := xml.MarshalIndent(people, "", "  ")
	return out
}

func main() {
	// Reads the proprietary people data format and turns it into XML
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString(0)

	people, err := stringToPeople(input)
	if err == nil {
		xml := peopleToXml(*people)
		fmt.Print(string(xml))
	} else {
		fmt.Println(err)
		os.Exit(1)
	}
}

File: python/convert.py
import sys
from xml.dom import minidom


class ParseError(Exception):

    def __init__(self, line, str):
        return super().__init__(f"Line {line}: {str}")


class Telephone:
    def __init__(self):
        self.mobile = None
        self.landline = None


class Address:
    def __init__(self):
        self.street = None
        self.city = None
        self.areacode = None


class Family:
    def __init__(self):
        self.name = None
        self.born = None
        self.telephone = None
        self.address = None


class Person:
    def __init__(self):
        self.family = []
        self.address = None
        self.first_name = None
        self.last_name = None
        self.telephone = None
        self.family = []


def create_xml_element(root, parent, name, dic):
    """ Creates an xml element under a parent with a name with attributes given
        in a supplied dictionary """

    element = root.createElement(name)
    parent.appendChild(element)

    for key, val in dic.items():
        if val is None:
            continue
        child_ele = root.createElement(key)
        element.appendChild(child_ele)
        child_ele.appendChild(root.createTextNode(val))

    return element


def convert_to_xml(input):
    """ The main function that converts a list of people to XML format """

    current_person = None
    current_family = None
    line = 1
    persons = []

    for s in input.split('\n'):

        args = s.strip().split('|')
        if args[0] == 'P': # Person
            if len(args) != 3:
                raise ParseError(line, "P: not enough arguments")
            
            _, first_name, last_name = args
            current_person = Person()
            current_person.first_name = first_name
            current_person.last_name = last_name
            current_family = None

            persons.append(current_person)

        if args[0] == 'F': # Family
            if not current_person:
                raise ParseError(line, "F: no current person")

            if len(args) != 3:
                raise ParseError(line, "F: not enough arguments")
            
            _, name, birth = args
            current_family = Family()
            current_family.name = name
            current_family.born = birth
            current_person.family.append(current_family)

        if args[0] == 'T': # Telephone, assigns to the current person or current family [member]
            if not current_person and not current_family:
                raise ParseError(line, "T: no current person or family")

            if len(args) != 3:
                raise ParseError(line, "T: not enough arguments")
            
            _, mobile, landline = args
            # Assign to the current family if they have no telephone, otherwise this is repeated and we assign to the person
            phone = (current_family if current_family and not current_family.telephone else current_person).telephone = Telephone()
            phone.mobile = mobile
            phone.landline = landline

        if args[0] == 'A': # Address, assigns to the current person or current family
            if not current_person and not current_family:
                raise ParseError(line, "A: no current person or family")

            if len(args) != 4:
                raise ParseError(line, "A: not enough arguments")
            
            _, street, city, areacode = args
            # Works the same as assigning telephone above
            address = (current_family if current_family and not current_family.address else current_person).address = Address()
            address.street = street
            address.city = city
            address.areacode = areacode

        line += 1

    root = minidom.Document()

    people = root.createElement('people') 
    root.appendChild(people)

    # Generate the actual xml
    for person in persons:

        person_element = create_xml_element(root, people, 'person', {
            "firstname": person.first_name,
            "lastname": person.last_name
        })

        if person.address:
            create_xml_element(root, person_element, 'address', {
                    "street": person.address.street,
                    "city": person.address.city,
                    "areacode": person.address.areacode
                }
            )

        if person.telephone:
            create_xml_element(root, person_element, 'phone', {
                "mobile": person.telephone.mobile,
                "landline": person.telephone.landline
            })

        
        for family in person.family:
            family_element = create_xml_element(root, person_element, 'family', {
                    "name": family.name,
                    "born": family.born
                }
            )

            if family.address:
                create_xml_element(root, family_element, 'address', {
                        "street": family.address.street,
                        "city": family.address.city,
                        "areacode": family.address.areacode
                    }
                )

            if family.telephone:
                create_xml_element(root, family_element, 'phone', {
                        "mobile": family.telephone.mobile,
                        "landline": family.telephone.landline
                    }
                )

    xml_str = people.toprettyxml(indent ="  ")

    return xml_str


def main():
    """ Reads the input format from stdin and prints out the corresponding XML to stdout """
    input = ""

    while True:
        s = sys.stdin.readline()
        if not s:
            break 
        input += s + '\n'

    xml_str = convert_to_xml(input)
    print(xml_str)


if __name__ == '__main__':
     main()
